tests/rptest/services/compatibility/kafka_streams_examples.py

```diff
@@ -171,3 +171,75 @@ def __init__(self, redpanda, is_driver):
 
         super(KafkaStreamsMapFunc, self).__init__(redpanda, is_driver,
                                                   jar_args)
+
+
+class KafkaStreamsInteractiveQueries(KafkaStreams):
+    def __init__(self, redpanda, is_driver, port):
+        jar_args = (
+            f"interactivequeries.WordCountInteractiveQueriesDriver {redpanda.brokers()}",
+            f"interactivequeries.WordCountInteractiveQueriesExample {port} {redpanda.brokers()}"
+        )
+
+        super(KafkaStreamsInteractiveQueries,
+              self).__init__(redpanda, is_driver, jar_args)
+
+        self._is_driver = is_driver
+        self._node_name = None
+
+    def driver_cond(self, line):
+        return "Driver started." in line
+
+    def set_node_name(self, node_name):
+        # The non-driver also needs the node name
+        # for the REST server
+        if not self._is_driver:
+            self._jar_arg += f" {node_name}"
+
+        self._node_name = node_name
+
+
+# Not use KafkaStreams base class here because this KStreams example
+# uses a different maven phase and args to run
+class KafkaStreamsMicroservices(ExampleBase):
+    def __init__(self, redpanda, jar):
+        super(KafkaStreamsMicroservices, self).__init__(redpanda)
+
+        self._jar = jar
+
+    def _condition(self, line):
+        return "BUILD SUCCESS" in line
+
+    # Return the command to call in the shell
+    def cmd(self):
+        cmd = f"mvn compile -Dtest=io.confluent.examples.streams.microservices.{self._jar}"
+        cmd = cmd + f" test -f {TESTS_DIR}/pom.xml"
+        return cmd
+
+    # Return the process name to kill
+    def process_to_kill(self):
+        return "mvn"
+
+
+class KafkaStreamsAppReset(KafkaStreams):
```

Comment:
The problem with this code is that inheritance scatters logic across the whole file (files) and make it easy to overlook potential issues during review and overall makes it harder to read and to navigate through the code.

In order to understand what `KafkaStreamsAppReset` does I had to jump through `KafkaStreams`, `ExampleBase` and look into the other classes to understand what `is_driver` means. Sometime duplication is better than an abstraction and in this case it seems we may get away even without the duplication. First, let's collapse the `KafkaStreamsAppReset`'s inheritance tree into a single class:

```(python)
class KafkaStreamsAppReset:
    def __init__(self, redpanda):
        self._condition_met = False
        self._redpanda = redpanda
    
    def condition(self, line):
        self._condition_met = "Example started." in line
    
    def condition_met(self):
        return self._condition_met
    
    def set_node_name(self, node_name):
        pass

    def cmd(self):
        brokers = self._redpanda.brokers()
        cmd = f"java -cp /opt/kafka-streams-examples/target/kafka-streams-examples-6.2.0-standalone.jar io.confluent.examples.streams.ApplicationResetExample {brokers} --reset"
        return cmd

    def process_to_kill(self):
        return "java"
```

It isn't much longer that the original one but all the logic is concentrated in one place. But we may reduce it even further. If
  
  - example_runner passed `node_name` as cmd's argument
  - example_runner internally tracks if a condition is met

when the code collapses to:

```(python)
class KafkaStreamsAppReset:
    def __init__(self, redpanda):
        self._redpanda = redpanda

    def condition(self, line):
        return "Example started." in line
    
    def cmd(self, node_name):
        brokers = self._redpanda.brokers()
        cmd = f"java -cp /opt/kafka-streams-examples/target/kafka-streams-examples-6.2.0-standalone.jar io.confluent.examples.streams.ApplicationResetExample {brokers} --reset"
        return cmd
    
    def process_to_kill(self):
        return "java"
```

Which just a few lines longer that the original. What do you think?

Reply:
I see your point. I had a similar structure a while ago but decided to generalize repeated code into parent classes after several recommendations from others.

I'll expand the classes in the next push.